File: app/src/main/java/org/witness/proofmode/onboarding/OnboardingActivity.java
package org.witness.proofmode.onboarding;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.os.Handler;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageButton;

import com.github.paolorotolo.appintro.AppIntroFragment;

import org.witness.proofmode.MainActivity;
import org.witness.proofmode.R;
import org.witness.proofmode.SettingsActivity;
import org.witness.proofmode.onboarding.DottedProgressView;
import org.witness.proofmode.onboarding.NoSwipeViewPager;
import org.witness.proofmode.onboarding.OnboardingStepListener;
import org.witness.proofmode.onboarding.PrivacyFragment;
import org.witness.proofmode.onboarding.StepFragment;
import org.witness.proofmode.onboarding.WelcomeFragment;

import java.util.ArrayList;
import java.util.List;

public class OnboardingActivity extends AppCompatActivity implements OnboardingStepListener {

    // Set this arg to "true" to only show the tutorial steps
    public static final String ARG_ONLY_TUTORIAL = "only_tutorial";

    private NoSwipeViewPager pager;
    private DottedProgressView indicator;
    private ImageButton btnNext;
    private List<Fragment> fragmentList;
    private boolean onlyTutorial;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().hide();
        setContentView(R.layout.activity_onboarding);

        pager = findViewById(R.id.pager);
        indicator = findViewById(R.id.indicator);
        btnNext = findViewById(R.id.btnNext);

        fragmentList = new ArrayList<>();

        pager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int i, float v, int i1) {
            }

            @Override
            public void onPageSelected(int i) {
                if (onlyTutorial) {
                    indicator.setCurrentDot(i);
                } else {
                    indicator.setCurrentDot(i - 1);
                }
                showHideIndicator();
            }

            @Override
            public void onPageScrollStateChanged(int i) {

            }
        });

        btnNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onNextPressed();
            }
        });

        onlyTutorial = getIntent().getBooleanExtra(ARG_ONLY_TUTORIAL, false);

        if (!onlyTutorial) {
            fragmentList.add(WelcomeFragment.newInstance());
        }
        fragmentList.add(StepFragment.newInstance(R.string.onboarding_step1, R.string.onboarding_step1_title, R.string.onboarding_step1_content, R.drawable.ic_ill_tut01, 0));
        fragmentList.add(StepFragment.newInstance(R.string.onboarding_step2, R.string.onboarding_step2_title, R.string.onboarding_step2_content, R.drawable.ic_ill_tut02, 20));
        fragmentList.add(StepFragment.newInstance(R.string.onboarding_step3, R.string.onboarding_step3_title, R.string.onboarding_step3_content,R.drawable.ic_ill_tut03, 40));
        fragmentList.add(StepFragment.newInstance(R.string.onboarding_step4, R.string.onboarding_step4_title, R.string.onboarding_step4_content, R.drawable.ic_ill_tut04, 60));
        if (!onlyTutorial) {
            fragmentList.add(PrivacyFragment.newInstance());
        }

        // Set adapter
        if (onlyTutorial) {
            indicator.setNumberOfDots(fragmentList.size());
        } else {
            indicator.setNumberOfDots(fragmentList.size() - 2);
        }
        indicator.setCurrentDot(0);
        pager.setOffscreenPageLimit(fragmentList.size());
        pager.setAdapter(new OnboardingPagerAdapter(getSupportFragmentManager()));

        showHideIndicator();
    }

    private void showHideIndicator() {
        if (onlyTutorial) {
            indicator.setVisibility(View.VISIBLE);
            btnNext.setVisibility(View.VISIBLE);
            pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.all);
            return;
        }
        int lastItem = fragmentList.size() - 1;
        int currentItem = pager.getCurrentItem();
        indicator.setVisibility((currentItem == 0 || currentItem == lastItem) ? View.GONE : View.VISIBLE);
        btnNext.setVisibility((currentItem == 0 || currentItem == lastItem) ? View.GONE : View.VISIBLE);
        if (currentItem == 0) {
            pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.none);
        } else if (currentItem == 1) {
            pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.right);
        } else {
            pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.all);
        }
    }

    @Override
    public void onNextPressed() {
        if (pager.getCurrentItem() < (fragmentList.size() - 1)) {
            pager.setCurrentItem(pager.getCurrentItem() + 1, true);
        } else {
            // Done
            finish();
        }
    }

    @Override
    public void onPreviousPressed() {
        if (pager.getCurrentItem() > 0) {
            pager.setCurrentItem(pager.getCurrentItem() - 1, true);
        }
    }

    @Override
    public void onSettingsPressed() {
        Intent intent = new Intent(this, SettingsActivity.class);
        startActivity(intent);
        finish();
    }

    private class OnboardingPagerAdapter extends FragmentPagerAdapter {

        public OnboardingPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int i) {
            return fragmentList.get(i);
        }

        @Override
        public int getCount() {
            return fragmentList.size();
        }
    }
}

File: app/src/main/java/org/witness/proofmode/onboarding/OnboardingActivity.kt
package org.witness.proofmode.onboarding

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ImageButton
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import org.witness.proofmode.R
import org.witness.proofmode.SettingsActivity
import org.witness.proofmode.databinding.ActivityOnboardingBinding

class OnboardingActivity : AppCompatActivity(), OnboardingStepListener {
    private lateinit var binding: ActivityOnboardingBinding
    private lateinit var pager: NoSwipeViewPager
    private lateinit var indicator: DottedProgressView
    private lateinit var btnNext: ImageButton
    private var fragmentList: MutableList<Fragment> = ArrayList()
    private var onlyTutorial = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityOnboardingBinding.inflate(layoutInflater)

        supportActionBar?.hide()
        setContentView(binding.root)
        pager = binding.pager
        indicator = binding.indicator
        btnNext = binding.btnNext
        fragmentList = ArrayList()
        pager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(i: Int, v: Float, i1: Int) {}
            override fun onPageSelected(i: Int) {
                if (onlyTutorial) {
                    indicator.currentDot = i
                } else {
                    indicator.currentDot = i - 1
                }
                showHideIndicator()
            }

            override fun onPageScrollStateChanged(i: Int) {}
        })
        btnNext.setOnClickListener { onNextPressed() }
        onlyTutorial = intent.getBooleanExtra(ARG_ONLY_TUTORIAL, false)
        if (!onlyTutorial) {
            fragmentList.add(WelcomeFragment.newInstance())
        }
        fragmentList.add(
            StepFragment.newInstance(
                R.string.onboarding_step1,
                R.string.onboarding_step1_title,
                R.string.onboarding_step1_content,
                R.drawable.ic_ill_tut01,
                0
            )
        )
        fragmentList.add(
            StepFragment.newInstance(
                R.string.onboarding_step2,
                R.string.onboarding_step2_title,
                R.string.onboarding_step2_content,
                R.drawable.ic_ill_tut02,
                20
            )
        )
        fragmentList.add(
            StepFragment.newInstance(
                R.string.onboarding_step3,
                R.string.onboarding_step3_title,
                R.string.onboarding_step3_content,
                R.drawable.ic_ill_tut03,
                40
            )
        )
        fragmentList.add(
            StepFragment.newInstance(
                R.string.onboarding_step4,
                R.string.onboarding_step4_title,
                R.string.onboarding_step4_content,
                R.drawable.ic_ill_tut04,
                60
            )
        )
        if (!onlyTutorial) {
            fragmentList.add(PrivacyFragment.newInstance())
        }

        // Set adapter
        if (onlyTutorial) {
            indicator.numberOfDots = fragmentList.size
        } else {
            indicator.numberOfDots = fragmentList.size - 2
        }
        indicator.currentDot = 0
        pager.offscreenPageLimit = fragmentList.size
        pager.adapter = OnboardingPagerAdapter(supportFragmentManager)
        showHideIndicator()
    }

    private fun showHideIndicator() {
        if (onlyTutorial) {
            indicator.visibility = View.VISIBLE
            btnNext.visibility = View.VISIBLE
            pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.ALL)
            return
        }
        val lastItem = fragmentList.size - 1
        val currentItem = pager.currentItem
        indicator.visibility =
            if (currentItem == 0 || currentItem == lastItem) View.GONE else View.VISIBLE
        btnNext.visibility =
            if (currentItem == 0 || currentItem == lastItem) View.GONE else View.VISIBLE
        when (currentItem) {
            0 -> {
                pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.NONE)
            }

            1 -> {
                pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.RIGHT)
            }

            else -> {
                pager.setAllowedSwipeDirection(NoSwipeViewPager.SwipeDirection.ALL)
            }
        }
    }

    override fun onNextPressed() {
        if (pager.currentItem < fragmentList.size - 1) {
            pager.setCurrentItem(pager.currentItem + 1, true)
        } else {
            // Done
            finish()
        }
    }

    override fun onPreviousPressed() {
        if (pager.currentItem > 0) {
            pager.setCurrentItem(pager.currentItem - 1, true)
        } else if (onlyTutorial) {
            // Close
            finish()
        }
    }

    override fun onBackPressed() {
        if (pager.currentItem > 0) {
            onPreviousPressed()
            return
        }
        super.onBackPressed()
    }

    override fun onSettingsPressed() {
        val intent = Intent(this, SettingsActivity::class.java)
        startActivity(intent)
    }

    private inner class OnboardingPagerAdapter(fm: FragmentManager?) : FragmentPagerAdapter(
        fm!!
    ) {
        override fun getItem(i: Int): Fragment {
            return fragmentList[i]
        }

        override fun getCount(): Int {
            return fragmentList.size
        }
    }

    companion object {
        // Set this arg to "true" to only show the tutorial steps
        const val ARG_ONLY_TUTORIAL = "only_tutorial"
    }
}